x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -133,32 +140,73 @@ public synchronized void closeSession(String sessionUUID) {
         IOUtils.closeWhileHandlingException(restore);
     }
 
+    public synchronized Reader getSessionReader(String sessionUUID, String fileName) throws IOException {
+        RestoreContext restore = onGoingRestores.get(sessionUUID);
+        if (restore == null) {
+            logger.info("could not get session [{}] because session not found", sessionUUID);
+            throw new IllegalArgumentException("session [" + sessionUUID + "] not found");
+        }
+        return restore.getFileReader(fileName);
```

Comment:
this means that opening the index input is under a global lock. If there are multiple shards recovering from remote, this might become a performance bottleneck.

Reply:
I made some of the changes we discussed.